jlm/llvm/opt/alias-analyses/OnlineCycleDetection.hpp

```diff
@@ -0,0 +1,368 @@
+    // All nodes not visited by the DFS are compacted to the left,
+    // All other nodes are added in order in the supersetAndBeyond list
+    // Nodes that are part of the cycle, if any, are ignored
+    int64_t nextTopologicalIndex = lowerTopo;
+    for (int64_t i = lowerTopo; i <= upperTopo; i++)
+    {
+      // Skip all topological indices that do not contain a node
+      if (TopoOrderToObject_[i] == -1)
+        continue;
+
+      PointerObjectIndex node = TopoOrderToObject_[i];
+      if (DfsCycleNodes_.Contains(node))
+        continue;
+
+      JLM_ASSERT(ObjectToTopoOrder_[node] == i);
+
+      if (DfsNodesVisited_.Contains(node))
+      {
+        DfsSupersetAndBeyond_.push_back(node);
+        continue;
+      }
+
+      // Add node to its new position
+      ObjectToTopoOrder_[node] = nextTopologicalIndex;
+      TopoOrderToObject_[nextTopologicalIndex] = node;
+      nextTopologicalIndex++;
+    }
+
+    // Next, add all elements from OCD_supersetAndBeyond
+    for (auto node : DfsSupersetAndBeyond_)
+    {
+      JLM_ASSERT(Set_.IsUnificationRoot(node));
+
+      ObjectToTopoOrder_[node] = nextTopologicalIndex;
+      TopoOrderToObject_[nextTopologicalIndex] = node;
+      nextTopologicalIndex++;
+    }
+
+    // Any leftover positions in the topological order should now become unoccupied
+    while (nextTopologicalIndex <= upperTopo)
+    {
+      TopoOrderToObject_[nextTopologicalIndex] = -1;
+      nextTopologicalIndex++;
+    }
+
+#ifdef JLM_ENABLE_ASSERTS
+    ValidateTopologicalOrder();
+#endif
+
+    // return the root of the unification if a cycle was detected, otherwise nullopt
+    return optUnificationRoot;
+  };
+
+  /**
+   * Function performing asserts to ensure all invariants are maintained.
+   * Nodes should have a position in the topological order iff they are unification roots.
+   * All successor roots should come later in the topological order.
+   * The bi-directional lookup, node <-> topological index, is maintained.
+   */
+  void
+  ValidateTopologicalOrder()
+  {
+    for (PointerObjectIndex node = 0; node < Set_.NumPointerObjects(); node++)
+    {
+      // Non-unification roots should not have a place in the topological order
+      if (!Set_.IsUnificationRoot(node))
+      {
+        JLM_ASSERT(ObjectToTopoOrder_[node] == -1);
+        continue;
+      }
+
+      // Ensure we have a position in the topological order
+      const auto topo = ObjectToTopoOrder_[node];
+      JLM_ASSERT(topo != -1 && TopoOrderToObject_[topo] == node);
+
+      // Ensure all outgoing edges go to unifications with higher topological index
+      for (auto successor : GetSuccessors_(node))
+      {
+        const auto successorRoot = Set_.GetUnificationRoot(successor);
+        if (successorRoot == node) // Ignore self-edges
+          continue;
+        JLM_ASSERT(ObjectToTopoOrder_[successorRoot] != -1);
+        JLM_ASSERT(ObjectToTopoOrder_[successorRoot] > topo);
```

Comment:
This is now the second or third time I see this little code snippet. Could you introduce a little method that takes two parameters and checks these invariants, and then use this here with `JLM_ASSERT()`?